pkg/accounting/accounting.go

```diff
@@ -0,0 +1,210 @@
+// license that can be found in the LICENSE file.
+
+package accounting
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+var _ Interface = (*Accounting)(nil)
+
+// Interface is the main interface for Accounting
+type Interface interface {
+	// Reserve reserves a portion of the balance for peer
+	// It returns an error if the operation would risk exceeding the disconnect threshold
+	// This should be called (always in combination with release) before a Credit action
+	Reserve(peer swarm.Address, price uint64) error
+	// Release releases reserved funds
+	Release(peer swarm.Address, price uint64)
+	// Credit increases the balance the peer has with us (we "pay" the peer)
+	Credit(peer swarm.Address, price uint64) error
+	// Debit increases the balance we have with the peer (we get "paid")
+	Debit(peer swarm.Address, price uint64) error
+	// Balance returns the current balance for the given peer
+	Balance(peer swarm.Address) (int64, error)
+}
+
+// PeerBalance holds all relevant accounting information for one peer
+type PeerBalance struct {
+	lock     sync.Mutex
+	balance  int64  // amount that the peer owes us if positive, our debt if negative
+	reserved uint64 // amount currently reserved for active peer interaction
+}
+
+// Options for accounting
+type Options struct {
+	DisconnectThreshold uint64
+	Logger              logging.Logger
+	Store               storage.StateStorer
+}
+
+// Accounting is the main implementation of the accounting interface
+type Accounting struct {
+	balancesMu sync.Mutex // mutex for accessing the balances map
+	balances   map[string]*PeerBalance
+	logger     logging.Logger
+	store      storage.StateStorer
+
+	disconnectThreshold uint64 // the debt threshold at which we will disconnect from a peer
+}
+
+// NewAccounting creates a new Accounting instance with the provided options
+func NewAccounting(o Options) *Accounting {
+	return &Accounting{
+		balances:            make(map[string]*PeerBalance),
+		disconnectThreshold: o.DisconnectThreshold,
+		logger:              o.Logger,
+		store:               o.Store,
+	}
+}
+
+// Reserve reserves a portion of the balance for peer
+func (a *Accounting) Reserve(peer swarm.Address, price uint64) error {
+	balance, err := a.getPeerBalance(peer)
+	if err != nil {
+		return err
+	}
+
+	balance.lock.Lock()
+	defer balance.lock.Unlock()
+
+	// the previously reserved balance plus the new price is the maximum amount paid if all current operations are successful
+	// since we pay this we have to reduce this (positive quantity) from the balance
+	// the disconnectThreshold is stored as a positive value which is why it must be negated prior to comparison
+	if balance.balance-int64(balance.reserved+price) < -int64(a.disconnectThreshold) {
+		return fmt.Errorf("cannot afford operation with peer %v", peer.String())
+	}
```

Comment:
This is not supposed to be the final version of this check. When all is done this check should be done with the peers payment threshold (plus some tolerance) which we will know either through some handshake or other protocol messages. But the payment threshold does not exist yet (in this iteration of accounting, that would be the next sprint) so I thought the disconnect threshold (which used to be hardcoded until recently and we could therefore assume that another well-behaving peer would have the same) would be a good alternative in the meantime. But since `disconnectThreshold` is now configurable, maybe we should remove the check until the notion of payment threshold is reintroduced.

Reply:
In both cases (removing or not removing) a TODO would be a good idea, mentioning that there must be a check on the payment threshold.
I think removing + TODO would be best. Leaving it up to you!

Reply:
Hm, I was about to remove this but the problem is that once this check is gone I cannot test the `Reserve` / `Release` functionality at all (because accounting does not expose the internals of that). Perhaps it's better to leave that in and just assume a fixed (non-configurable) payment threshold at least for now.

Reply:
Alternatively introduce an interface
```go
type PaymentThresholder interface {
  PaymentThreshold(peer swarm.Address) uint64
}
```
and have a fixed value implementation of this injected into `Accounting`. Later the accounting code wouldn't even need changing if the price exchange protocol implements that interface.